File: packages/rust-fizzbuzz/src/arithmetic.rs
use crate::types::Arithmetic;

pub struct StandardArithmetic;

impl Arithmetic<i32> for StandardArithmetic {
    fn is_divisible(&self, n: i32, divisor: i32) -> bool {
        n % divisor == 0
    }
    fn to_string(&self, n: i32) -> String {
        n.to_string()
    }
}

File: packages/core-logic/src/arithmetic.ts
import { Arithmetic } from './types.js';

/**
 * ADR 009: Unbounded BigInt arithmetic.
 */
export class BigIntArithmetic implements Arithmetic<bigint> {
  isDivisible(n: bigint, divisor: bigint): boolean {
    return n % divisor === 0n;
  }
  modulo(n: bigint, divisor: bigint): bigint {
    return n % divisor;
  }
  toString(n: bigint): string {
    return n.toString();
  }
}

/**
 * ADR 009: Checked 64-bit machine arithmetic.
 */
export class CheckedArithmetic implements Arithmetic<number> {
  isDivisible(n: number, divisor: number): boolean {
    if (!Number.isSafeInteger(n)) {
      throw new Error(`Precision loss detected: ${n} is not a safe integer`);
    }
    return n % divisor === 0;
  }
  modulo(n: number, divisor: number): number {
    return n % divisor;
  }
  toString(n: number): string {
    return n.toString();
  }
}
